app/controllers/application_controller.rb

```diff
@@ -3,8 +3,139 @@
 class ApplicationController < Sinatra::Base
 
   configure do
+    enable :sessions
+    set :session_secret, 'billy'
     set :public_folder, 'public'
     set :views, 'app/views'
   end
 
-end
\ No newline at end of file
+  get '/' do
+    erb :index
+  end
+
+  get '/signup' do
+    if session[:id]
+      redirect '/tweets'
+    end
+      erb :signup
+  end
+
+  post '/signup' do
+    @user = User.find_or_create_by(username: params['username'], email: params['email'], password: params['password'])
+    if @user.save
+      session[:id] = @user.id
+      redirect '/tweets'
+    else
+      redirect '/signup'
+    end
+  end
+
+  get '/login' do
+    if session[:id] == nil
+      erb :login
+    else
+      redirect '/tweets'
+    end
+  end
+
+  post '/login' do
+    @user = User.find_by(username: params['username'], password: params['password'])
+    session[:id] = @user.id
+    redirect '/tweets'
+  end
+
+  get '/tweets' do
```

Comment:
This controller is overloaded. Consider separating your concerns by splitting some of these actions into separate controllers (i.e. a users controller and a tweets controller).